lib/Macros/Sources/SwiftMacros/DebugDescriptionMacro.swift

```diff
@@ -0,0 +1,312 @@
+// `a.b.c` will result in callbacks for each DeclReference `a`, `b`, and then `c`.
+private func enumerateMembers(_ memberAccess: MemberAccessExprSyntax, _ action: (DeclReferenceExprSyntax) -> Void) {
+    if let baseMember = memberAccess.base?.as(MemberAccessExprSyntax.self) {
+        enumerateMembers(baseMember, action)
+    } else if let baseDecl = memberAccess.base?.as(DeclReferenceExprSyntax.self) {
+        action(baseDecl)
+    }
+    action(memberAccess.declName)
+}
+
+// MARK: - Encoding
+
+/// Construct an LLDB type summary record.
+///
+/// The record is serializeed as a tuple of `UInt8` bytes.
+///
+/// The record contains the following:
+///   * Version number of the record format
+///   * The size of the record (encoded as ULEB)
+///   * The type identifier, which is either a type name, or for generic types a type regex
+///   * The description string converted to an LLDB summary string
+///
+/// The strings (type identifier and summary) are encoded with both a length prefix (also ULEB)
+/// and with a null terminator.
+private func encodeTypeSummaryRecord(_ typeIdentifier: String, _ summaryString: String) -> String {
+    let encodedType = encodeString(typeIdentifier)
+    let encodedSummary = encodeString(summaryString)
+    let recordSize = UInt(encodedType.count + encodedSummary.count)
+    return """
+    /* version */ 1 as UInt8,
+    /* record size */ \(literalBytes(encodeULEB(recordSize))),
+    /* "\(typeIdentifier)" */ \(literalBytes(encodedType)),
+    /* "\(summaryString)" */ \(literalBytes(encodedSummary))
+    """
+}
+
+/// Generate a _partial_ Swift literal from the given bytes. It is partial in that must be embedded
+/// into some other syntax, specifically as a tuple.
+private func literalBytes(_ bytes: [UInt8]) -> String {
+    bytes.map({ "\($0) as UInt8" }).joined(separator: ", ")
+}
+
+/// Encode a string into UTF8 bytes, prefixed by a ULEB length, and suffixed by the null terminator.
+private func encodeString(_ string: String) -> [UInt8] {
+    let size = UInt(string.utf8.count) + 1 // including null terminator
+    var bytes: [UInt8] = []
+    bytes.append(contentsOf: encodeULEB(size))
+    bytes.append(contentsOf: string.utf8)
+    bytes.append(0) // null terminator
+    return bytes
+}
+
+/// Encode an unsigned integer into ULEB format. See https://en.wikipedia.org/wiki/LEB128
+private func encodeULEB(_ value: UInt) -> [UInt8] {
+    guard value > 0 else {
+        return [0]
+    }
+
+    var bytes: [UInt8] = []
+    var buffer = value
+    while buffer > 0 {
+        var byte = UInt8(buffer & 0b01111111)
+        buffer >>= 7
+        if buffer > 0 {
+            byte |= 0b10000000
+        }
+        bytes.append(byte)
+    }
+    return bytes
+}
+
+// MARK: - Extensions
+
+extension Collection {
+    /// Convert a single element collection to a single value. When a collection consists of
+    /// multiple elements, nil is returned.
+    fileprivate var single: Element? {
```

Comment:
The standard term that we use for this is `only`. Don’t start searching for it, these extensions are defined everywhere.

For example

https://github.com/apple/swift-syntax/blob/52859074c793a46a0446e8b225686a0f5ce19d97/Sources/SwiftParserDiagnostics/Utils.swift#L46-L55